File: LeetCode/Medium/0053-maximum-subarray/0053-maximum-subarray.js
/**
 * @param {number[]} nums
 * @return {number}
 */
var maxSubArray = function (nums) {
    // let cur = nums[0];  // dp[i-1]：以当前元素结尾的最大子数组和
    // let ans = nums[0];  // 全局最大

    // for (let i = 1; i < nums.length; i++) {
    //     // 状态转移方程
    //     cur = Math.max(nums[i], cur + nums[i]);
    //     ans = Math.max(ans, cur);
    // }

    // return ans;

    //前缀和 + 最小前缀和版
    if (nums.length === 0) return 0;

    let ans = -Infinity;
    let curPrefixSum = 0;
    let minPrefixSum = 0;  // 相当于 prefix[-1] = 0

    for (const num of nums) {
        curPrefixSum += num; // 当前前缀和
        ans = Math.max(ans, curPrefixSum - minPrefixSum); // 尝试以当前位置结尾的最大子数组
        minPrefixSum = Math.min(minPrefixSum, curPrefixSum); // 维护历史最小前缀和
    }

    return ans;
};

File: LeetCode/Medium/0053-maximum-subarray/0053-maximum-subarray.cpp
class Solution {
public:
    // int maxSubArray(vector<int>& nums) {
    //     if (nums.empty()) {
    //         return 0;
    //     }
    //     int ans = INT_MIN;
    //     int curPrefixsum = 0;
    //     int minPrefixSum = 0;
    //     for (int num : nums) {
    //         curPrefixsum += num;
    //         ans = max(ans, curPrefixsum - minPrefixSum);
    //         minPrefixSum = min(minPrefixSum, curPrefixsum);
    //     }
    //     return ans;
    // }

    //     int maxSubArray(vector<int>& nums) {
    //     int n = (int)nums.size();
    //     vector<int> dp(n);
    //     dp[0] = nums[0];
    //     int ans = dp[0];
    //     for (int i = 1; i < n; ++i) {
    //         dp[i] = max(nums[i], dp[i - 1] + nums[i]);
    //         ans = max(ans, dp[i]);
    //     }
    //     return ans;
    // }

    // int maxSubArray(vector<int>& nums) {
    //     int cur = nums[0]; // 相当于 dp[i-1]
    //     int res = nums[0];
    //     for (int i = 1; i < (int)nums.size(); ++i) {
    //         cur = max(nums[i], cur + nums[i]); // dp[i]
    //         res = max(res, cur);
    //     }
    //     return res;
    // }

    int maxSubArray(vector<int>& nums) {
        int n = nums.size();
        if (n == 0)
            return 0;
        vector<int> dp(n);
        dp[0] = nums[0];
        int res = dp[0];

        for (int i = 1; i < n; i++) {
            dp[i] = max(nums[i], dp[i - 1] + nums[i]);
            res = max(res, dp[i]);
        }
        return res;
    }
};